### PR TITLE
EDU-428 Add favorites tab to dashboard

### PR DESCRIPTION
**Pull request by adrianaluchian**
Closes https://educandu.atlassian.net/browse/EDU-428

and looks like

Screen Shot 2022-03-11 at 12 12 30

### CODE REVIEW

File: src/server/dashboard-controller.js [hunk 1]
```diff
@@ -24,11 +24,14 @@ class UserController {
     if (this.serverConfig.areRoomsEnabled) {
       rooms = await this.roomService.getRoomsOwnedOrJoinedByUser(user._id);
     }
-    const mappedRooms = await Promise.all(rooms.map(room => this.clientDataMappingService.mapRoom(room, user)));
     const activities = await this.dashboardService.getUserActivities({ userId: user._id, limit: 10 });
+    const favorites = await this.dashboardService.getUserFavorites(user._id);
```
Comment (ahelmberger) [inline src/server/dashboard-controller.js:28]: I don't think we have to fetch them from the DB, they should be there by `user.favorites` (as the `req.user` is the DB user object).

File: src/services/dashboard-service.js [hunk 2]
```diff
@@ -136,6 +136,34 @@ class DashboardService {
 
     return activities;
   }
+
+  async getUserFavorites(userId) {
```
Comment (ahelmberger) [inline src/services/dashboard-service.js:140]: We could make this the user object itself (which is there in `req.user` when calling this service).

File: src/services/dashboard-service.js [hunk 3]
```diff
@@ -136,6 +136,34 @@ class DashboardService {
...
+    const user = await this.userStore.getUserById(userId);
+
+    const documentIds = user.favorites.filter(f => f.type === FAVORITE_TYPE.document).map(d => d.id);
+    const roomIds = user.favorites.filter(f => f.type === FAVORITE_TYPE.room).map(r => r.id);
+    const lessonIds = user.favorites.filter(f => f.type === FAVORITE_TYPE.lesson).map(l => l.id);
+
+    const documents = documentIds.length ? await this.documentStore.getDocumentsMetadataByKeys(documentIds) : [];
+    const rooms = roomIds.length ? await this.roomStore.getRoomsByIds(roomIds) : [];
+    const lessons = lessonIds.length ? await this.lessonStore.getLessonsMetadataByIds(lessonIds) : [];
```
Comment (ahelmberger) [inline src/services/dashboard-service.js:149]: We could parallelize this by `Promise.all`

File: src/stores/document-store.js [hunk 4]
```diff
@@ -54,6 +54,10 @@ class DocumentStore {
     return this.collection.distinct('key');
   }
 
+  getDocumentsMetadataByKeys(keys, { session } = {}) {
+    return this.collection.find({ key: { $in: keys } }, { session }).toArray();
```
Comment (ahelmberger) [inline src/stores/document-store.js:58]: There is no index on the `key` in documents, only on `_id` (which has the same value)

File: src/services/dashboard-service.js (hunk 3 above)
Comment (adrianaluchian) [reply src/services/dashboard-service.js]: Indeed, done, thanks!

File: src/services/dashboard-service.js (hunk 2 above)
Comment (adrianaluchian) [reply src/services/dashboard-service.js]: True, cause the `user` in the controller has the whole data 🤔 
Changed, thanks.

File: src/stores/document-store.js (hunk 4 above)
Comment (adrianaluchian) [reply src/stores/document-store.js:58]: Funny, did not realize that documents don't have a `key` index. I agree, maybe we can get around to removing this field eventually. Till then I decided to add the index, to make sure we don't suffer from this.

File: src/services/dashboard-service.js (hunk 3 above)
Comment (ahelmberger) [reply src/services/dashboard-service.js]: :+1:

**Review by ahelmberger — APPROVED** [review]

File: src/stores/document-store.js (hunk 4 above)
Comment (ahelmberger) [reply src/stores/document-store.js:58]: :+1:

**Pull request merged**